### PR TITLE
AutoDiff associated functions in witness table, using SILDeclRef

### PR DESCRIPTION
**Pull request by marcrasi**
This deletes the `AutoDiffAssociatedFunctionWitness` from the witness table, and instead puts the autodiff associated functions in normal method witness table entries, using the autodiff SILDeclRef from https://github.com/apple/swift/pull/21224.

For easier reviewing, I have split it into two commits. The first commit removes most of the `AutoDiffAssociatedFunctionWitness` stuff, putting everything back mostly like it was in `master`, and the second commit adds the entries back in, using SILDeclRef.

cc @slavapestov @rjmccall

### CODE REVIEW

File: include/swift/SIL/SILDeclRef.h [hunk 1]
```diff
@@ -361,6 +361,13 @@ struct SILDeclRef {
     return r;
   }
 
+  /// Returns this `SILDeclRef` with the `loc` replaced with `decl`.
```
Comment (rxwei) [inline include/swift/SIL/SILDeclRef.h]: Should this be SWIFT_ENABLE_TENSORFLOW?

**Review by rxwei — APPROVED** [review]
Looks like it's hugely simplified!

File: include/swift/SIL/SILDeclRef.h (hunk 1 above)
Comment (marcrasi) [reply include/swift/SIL/SILDeclRef.h]: The pieces above it are tensorflow too, and they have a SWIFT_ENABLE_TENSORFLOW.

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux CPU

**Pull request merged**